pam_authz/pam/display.c

```diff
@@ -0,0 +1,203 @@
+static const char DISPLAY_STYLE_TEXT_INFO[]       = "info";
+static const char DISPLAY_STYLE_ERROR_MSG[]       = "error";
+
+
+static const struct DisplayStyleToPamInt {
+	const char *style;
+	const int  pam_int;
+} DISPLAY_STYLE_TO_PAM_INT[DISPLAY_STYLE_COUNT] = {
+	{DISPLAY_STYLE_PROMPT_ECHO_ON,  PAM_PROMPT_ECHO_ON},
+	{DISPLAY_STYLE_PROMPT_ECHO_OFF, PAM_PROMPT_ECHO_OFF},
+	{DISPLAY_STYLE_ERROR_MSG,       PAM_ERROR_MSG},
+	{DISPLAY_STYLE_TEXT_INFO,       PAM_TEXT_INFO},
+};
+
+static int pam_int_for_display_style(const char *style) {
+	int i;
+	for (i = 0; i < DISPLAY_STYLE_COUNT; i++) {
+		if (strcmp(DISPLAY_STYLE_TO_PAM_INT[i].style, style) == 0) {
+			return DISPLAY_STYLE_TO_PAM_INT[i].pam_int;
+		}
+	}
+
+	return DISPLAY_STYLE_INVALID;
+}
+
+static struct pam_conv *get_pam_conv(pam_handle_t *pamh) {
+	if (!pamh)
+	return NULL;
+
+	struct pam_conv *conv;
+
+	pam_log(LOG_PRIORITY_DEBUG, "Retrieving struct pam_conv from the PAM handle.");
+
+	if (pam_get_item(pamh, PAM_CONV, (const void**)&conv) != PAM_SUCCESS)
+		return NULL;
+
+	return conv;
+}
+
+static char *call_pam_conv(pam_handle_t *pamh, int msg_style, char *message) {
+	// conv is always called with parameter num_msg == 1 as a way to
+	// make it compatible with both Linux-PAM and Solaris' PAM,
+	// see https://linux.die.net/man/3/pam_conv
+	int num_msg = 1;
+
+	// Create a struct pam_message array and populate it with a single object.
+	struct pam_message *msg_array = (struct pam_message *)(malloc(sizeof(struct pam_message)));
+	msg_array[0].msg_style = msg_style;
+	msg_array[0].msg = message;
+
+	// Create a struct pam_response array.
+	struct pam_response *resp_array = NULL;
+
+	struct pam_conv *obj = get_pam_conv(pamh);
+	if (obj == NULL)
+		return NULL;
+
+	pam_log(LOG_PRIORITY_DEBUG,
+		"Calling application-defined conversation function to display message '%s' to "
+		"the user.", message);
+
+	int conv_resp = obj->conv(num_msg,
+		(const struct pam_message **)&msg_array, &resp_array, obj->appdata_ptr);
+	if (conv_resp != PAM_SUCCESS) {
+		pam_log(LOG_PRIORITY_ERROR,
+			"Received error from application-defined conversation function: %s",
+			pam_strerror(pamh, conv_resp));
+
+		return NULL;
+	}
+
+	char *user_resp = NULL;
+	if (resp_array[0].resp != NULL) {
+		pam_log(LOG_PRIORITY_DEBUG, "Collected a prompt response from user.");
+
+		user_resp = strdup(resp_array[0].resp);
+		free((resp_array[0].resp));
+	}
+
+	free(resp_array);
+	free(msg_array);
+
+	return user_resp;
+}
+
+void
+engine_display(pam_handle_t *pamh, const char *endpoint,
+	struct DisplayResponses *display_responses_ptr) {
+	// Initialize empty responses, then fill it up as the user responses come in.
+	display_responses_ptr->count = 0;
+	// An empty malloc here allows calling free() later without having to check anything.
+	display_responses_ptr->responses = (struct DisplayResponse *)malloc(0);
+
+	json_t *result_j;
+	http_request(HTTP_METHOD_GET, endpoint, NULL, &result_j);
```

Comment:
What happens if the admin does not provide a display (or pull) policy to OPA? I think those should be optional.

In other words, if the display or pull decisions are empty or undefined, the PAM module should not return an error.

Reply:
The final return value of the PAM module depends solely on the response (if any) received from the authz policy. Lack of display/pull policies are logged.

Currently their absence does result in more error logs than necessary - despite the response from display/pull being empty, there will still be an attempt to unpack it, resulting in more error logs.

Reply:
OK, let's downgrade the log level to INFO or DEBUG or make sure the message communicates the fact that it's OK for the display and pull decisions are empty/undefined (not just "Unable to parse HTTP response JSON." because that indicates some kind of internal error occurred.)

Reply:
I added a check which will not attempt any of the display/pull stuff if those flags have not been provided, or have been given empty values.

The errors will still happen if the flags have been provided but cannot be acted upon.

Reply:
Requiring reconfiguration of the enforcement point because the policy changed is an anti-pattern. I think it's better to log the result at a lower level rather introduce a configuration flag.

Reply:
After realizing that it's easy to distinguish between user errors made while entering flags and policy structure errors in OPA, I've reduced the log severity of cases where display/pull is undefined to *info*.

Also added a note to the docs suggesting that it is best to specify all the policies once during PAM configuration, and then control all behavior via OPA in the future.